src/cyclebane/node_values.py

```diff
@@ -0,0 +1,302 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import abc
+from typing import TYPE_CHECKING, Any, Hashable, Iterable, Mapping, Sequence
+
+if TYPE_CHECKING:
+    import numpy
+    import pandas
+    import scipp
+    import xarray
+
+IndexName = Hashable
+IndexValue = Hashable
+
+
+class ValueArray(ABC):
+    """
+    Abstract base class for a series of values with an index that can be sliced.
+
+    Used by :py:class:`NodeValues` to store the values of a given node in a graph. The
+    abstraction allows for the use of different data structures to store the values of
+    nodes in a graph, such as pandas.DataFrame, xarray.DataArray, numpy.ndarray, or
+    simple Python iterables.
+    """
+
+    @staticmethod
+    def from_array_like(values: Any, *, axis_zero: int = 0) -> ValueArray:
+        if hasattr(values, 'dims'):
```

Comment:
It seems risky to rely on some arbitrary attribute name. Can't you use `isinstance` checks instead? (The same applies to `if hasattr(self._data_array, 'isel'):` and `if (columns := getattr(values, 'columns', None)) is not None:` below.)

Reply:
We cannot use `isinstance`, since the libraries might not be installed. I could use a string comparison on the class name instead?

Reply:
What about this?
```python
def _is_xarray_data_array(x: object) -> bool:
    try:
        import xarray as xr
    except ModuleNotFoundError:
        return False
    return isinstance(x, xr.DataArray)
```

Reply:
See update.